controller.go

```diff
@@ -469,3 +457,71 @@ func upgradePVFrom12(pv *PV) (deleted bool, err error) {
 	}
 	return false, nil
 }
+
+func provisionClaim(pvc *api.PersistentVolumeClaim) error {
+	// Provisioning can take some time. Start it in a separate goroutine.
+	// Keep a map of running provisioners and start new provisioning only if
+	// it's not already running. There will be some locks around, these are
+	// left out to implementation
+
+	if !runningProvisioners.isRunning(pvc) {
```

Comment:
For now, only one internal provisioner is allowed.

When we add VolumeSelectors and multiple provisioners, we should also define API for external provisioners. IMO we should somehow ensure that only single provisioner can provision a claim - otherwise two (or more) different provisioners will race to provision a volume and user may get different volume type each time he submits the same claim. I think that's bad user experience.


Note that current pseudocode copes with multiple provisioners quite gracefuly. Provisioned PV has deterministic name, only one provisioner succeeds creating it in etcd. Those who fail will delete the provisioned storage asset.